travis.sh

```diff
@@ -221,21 +225,24 @@ function prepare_ros_workspace() {
   run_script BEFORE_SCRIPT
 
   # For debugging: list the files in workspace's source folder
+  # Double check that we are in $ROS_WS/src after BEFORE_SCRIPt
   travis_run_simple cd $ROS_WS/src
   travis_run --title "List files in ROS workspace's source folder" ls --color=auto -alhF
 
   # Install source-based package dependencies
   travis_run --retry rosdep install -y -q -n --from-paths . --ignore-src --rosdistro $ROS_DISTRO
 
-  # Change to base of workspace
-  travis_run_simple cd $ROS_WS
-
   # Validate that we have some packages to build
   test -z "$(catkin list)" && echo -e "$(colorize RED Workspace $ROS_WS has no packages to build. Terminating.)" && exit 1
```

Comment:
There are exits and returns in many functions, which render your pushd/popd approach error-prone!
When returning from a function you would need to popd as well.